### ISSUE TITLE
recent libiconv upgrade broke install

### ISSUE THREAD

**Issue by its-pointless** [termux/termux-packages#3762]
**Problem description**
A clear and concise description of what the problem is. You may post screenshots in addition to description.

current upgrade leaves apt in a broken state since tar and coreutils don't upgrade properly leaving them not finding symbols in newer libiconv.
to fix a broken install

`apt remove coreutils tar `
and `apt remove iconv` if its installed 

then apt upgrade should work properly.

then
apt install coreutils tar 
should be fine

**Comment by Grimler91** [comment]
So `TERMUX_PKG_BREAKS="coreutils (<= 8.31-1), tar (<= 1.32-2)"` for libiconv might fix it? I'll give it a try

**Comment by Grimler91** [comment]
No, that doesn't work either it seems

**Comment by xeffyr** [comment]
Is `coreutils` or `tar` are not being upgraded when executing `pkg upgrade` ?

**Comment by Grimler91** [comment]
They get upgraded before libiconv is fully installed I suppose, so apt stops working because ˋrmˋ/ˋtarˋstops working due to missing  libiconv.so

**Comment by xeffyr** [comment]
@Grimler91 Perhaps apt should depend on libiconv ?

**Comment by Grimler91** [comment]
Maybe that would solve it indeed, I can try it tonight unless someone beats me to it

**Comment by xeffyr** [comment]
Restored my old Termux installation from backup, just to see order of package updates. Packages dpkg, gpgv, libcurl, liblzma, libnghttp2 are being updated before `apt`. So looks like apt updates its dependencies before itself.

Package `gpgv` doesn't have libiconv specified as dependency, so I guess issue comes from here.

**Comment by xeffyr** [comment]
> Package gpgv doesn't have libiconv specified as dependency, so I guess issue comes from here.

gpgv doesn't use/require libiconv so I will add it only for apt.

**Comment by xeffyr** [comment]
Adding libiconv to dependency of `apt` doesn't help much. Package libandroid-support will be installed before libiconv anyway and further installation process will be broken.

I see here only next 2 variants:
* Somehow add libiconv as dependency to libandroid-support. Will require fix for buildorder.py script to prevent infinite recursion caused by circular dependency.

* Introduce field "Pre-Depends". It described on https://www.debian.org/doc/debian-policy/ch-relationships.html:
  >This field is like Depends, except that it also forces dpkg to complete installation of the packages named before even starting the installation of the package which declares the pre-dependency

  >Pre-Depends should be used sparingly, preferably only by packages whose premature upgrade or installation would hamper the ability of the system to continue with any upgrade that might be in progress.

**Comment by xeffyr** [comment]
So:

* Adding libiconv as dependency will not solve this issue: libandroid-support will be removed during upgrade and symlink libiconv will disappear. Upgrade will stop because coreutils and tar no longer work.

* Pre-Depends will not work for same reason as above.

* libiconv as dependency for apt is useless. Again, because libandroid-support has to be upgraded before libiconv to prevent file conflicts.

* `TERMUX_PKG_BREAKS="coreutils (<= 8.31), tar (<= 1.32-1)"` will not do anything. Problem is in non-existent libiconv during upgrade.

**Comment by tomty89** [comment]
https://www.debian.org/doc/debian-policy/ch-relationships.html#packages-which-break-other-packages-breaks:
> When one binary package declares that it breaks another, dpkg will refuse to allow the package which declares Breaks to be unpacked unless the broken package is deconfigured first, and it will refuse to allow the broken package to be reconfigured.

I think a versioned breaks is the right approach, but it's not libiconv that's breaking the old coreutils and tar, but the new libandroid-support.

**Comment by jaredtlove** [comment]
Everything I've been sent to try it says no such directory found. 

Sent from Yahoo Mail on Android 
 
  On Tue, May 7, 2019 at 5:05 PM, tomty89<notifications@github.com> wrote:   
https://www.debian.org/doc/debian-policy/ch-relationships.html#packages-which-break-other-packages-breaks:


When one binary package declares that it breaks another, dpkg will refuse to allow the package which declares Breaks to be unpacked unless the broken package is deconfigured first, and it will refuse to allow the broken package to be reconfigured.


I think a versioned breaks is the right approach, but it's not libiconv that's breaking the old coreutils and tar, but the new libandroid-support.

**Comment by xeffyr** [comment]
> I think a versioned breaks is the right approach, but it's not libiconv that's breaking the old coreutils and tar, but the new libandroid-support.

No matter **old** coreutils or **new**. They both depend on libiconv.so.
If libiconv.so disappear, the dpkg will fail because commands like "rm" will throw linker errors.

**Comment by tomty89** [comment]
> Normally a Breaks entry will have an “earlier than” version clause; ... This use of Breaks will inform higher-level package management tools that the broken package must be upgraded before the new one.

So it will make sure the old coreutils and tar are upgraded before the old libandroid-support. With libiconv breaking libandroid-support, apt should be removing the old coreutils and tar, *and then* the old libandroid-support (*and then* install libiconv, *and then* the new coreutils, tar and libandroid-support).

I think the real problem is, the old coreutils and tar does not depends on libandroid-support (while links to a file of it). Therefore, apt *might* (apparently it does it in random order if there isn't a relationship between the packages? As I had no problem upgrading.) be removing libandroid-support first.

If that's the cause, we probably shouldn't even bother "fixing" it. We should just ask the users to perform a manual intervention (what @its-pointless has already proposed?).

**Comment by xeffyr** [comment]
Okay, I can now specify libandroid-support as "breaking" old coreutils and see if there problems.

**Comment by tomty89** [comment]
Note that it might not help coz it could be the "depends" in the old coreutils and tar that decides whether they always get removed before libandroid-support, which we cannot change (unless we time travel).

**Comment by xeffyr** [comment]
No, it doesn't help. It causes coreutils to be upgraded first which obviously won't solve problem.

**Comment by tomty89** [comment]
first? that's kind of vague.
It's expected that the old one could be removed before libandroid-support is (so it could be *removed* "first"), as they were not related (which cannot and probably should not be "fixed").
It is only *unexpected* that the new one could be installed before libiconv is installed. (It *should be* installed after the new libandroid-support is installed as well, because of https://github.com/termux/termux-packages/commit/9b72bfd4a282501cd6442021f7df6e38d976ee2c)

**Comment by xeffyr** [comment]
Yes, it printed something like unpacking coreutils 8.31-3 over 8.30-2.

**Comment by xeffyr** [comment]
Fixed with latest dpkg and libandroid-support.

**Comment by pedroapero** [comment]
Just got the same issue. I hadn't upgraded for a wile. I installed a random package (ncmpcpp) and I was surprised it made me install libandroid-support. Then my Termux was broken (same error for tar, ls, mv, cat etc…):
```
CANNOT LINK EXECUTABLE "mv": library "/data/data/com.termux/files/usr/lib/libiconv.so" not found
```

As you mentioned, I had to manually copy libiconv.so, coreutils and tar binaries (extracted from http://termux.net/dists/stable/main/binary-aarch64/). I used scp because cp/mv were broken.
Now everything upgraded, cheers!
